### PR TITLE
Invoke extensions before training

### PR DESCRIPTION
**Pull request by wkentaro**
## Why?

I sometimes need to know the loss and accuracy before the training (iteration=0), but currently I'm not sure how to do this with the Trainer class.
This PR achieves it.

## Demonstration

It writes loss and accuracy result even if the epoch is 0.

```
cd examples/mnist
python train_mnist.py
```

 

## ~Test~

```
cd tests/chainer_tests/training_tests/triggers_tests
python test_interval_trigger.py
```

### CODE REVIEW

**Comment by hvy** [comment]
Thank you for this PR. I definitely think that there is a demand for this.

Is it the case that you need to call the `__call__` method via `initialize` **in addition to** triggering the extension via `invoke_before_training`? In other words, is it not enough to call `initialize`?

**Comment by wkentaro** [comment]
@hvy 
Do you mean why do I need to change `IntervalTrigger`? It is because current `IntervalTrigger` doesn't fire at `iteration == 0`, so even if we call `__call__` method in `initialize`, that does not trigger the extension in actual. 

For example, `__call__` of `LogReport` is called in `initialize` in the MNIST example, but without `invoke_before_training=True` option to `IntervalTrigger`, below if-block is always skipped because it is the current behavior of `IntervalTrigger`.  
https://github.com/chainer/chainer/blob/1bc30968ddb9972b639f73e0686fe316be0e04b2/chainer/training/extensions/log_report.py#L78

**Comment by hvy** [comment]
Thank you for the explanation. I am not sure exactly which extensions you had in mind but I feel that it is better to actually implement the `initialize` methods for each of those to act appropriately. What do you think?

**Comment by wkentaro** [comment]
I think that is better. But the problem, in that case, is how to give users the option of choosing run `initialize` or not.
Currently, most of all extensions' `initialize` is None and it does not run by default. 

Previously, the extensions had an option of `invoke_before_training` for `__init__`, but it was removed after refactoring (maybe after v2?).
I am not sure whether we can re-add that option.

**Comment by hvy** [comment]
I don't think we should expect the user to set the `initialize` member with `__call__` in any case. It doesn't really speak for a good API. But I see that adding your third argument to the trigger is a flexible solution. @beam2d do you have any inputs on this?

**Comment by Hakuyume** [comment]
Modifying `IntervalTrigger` doesn't seem a good solution. I think each extension that can be called before training should implement `initialize`. If user can enable/disable `initialize`, it should be passed as an argument of the extension.

**Comment by hvy** [comment]
@Hakuyume thanks for the input. Could you explain why you think modifying the `IntervalTrigger` isn't a good solution?

**Comment by Hakuyume** [comment]
`IntervalTrigger` fires when `previous < n * interval <= current`. If we want `IntervalTrigger` fire at the first iterarion, it should be the first time the trigger called. This is not exactly same as `iteration == 0`.

For example, in the case that an `IntervalTrigger(4, 'iteration') is called at 3, 6 ,9, 12... iterations,
it should fire at 3 (0 * 4 <= 3), 6 (4 < 1 * 4 <= 6), 9 (6 < 2 * 4 <= 9), 12 (3 * 4 <=12) .

**Comment by hvy** [comment]
I think we need to reconsider if this is a applies to all extensions, a subset of them, or if firing at 0 should be the expected default behavior even for all of the extensions. If this applies to a subset, I feel https://github.com/chainer/chainer/pull/3511#issuecomment-350159097 is fine.

What do you think @wkentaro @Hakuyume ?

**Comment by wkentaro** [comment]
Thank you for the suggestion @Hakuyume @hvy . I agree with your idea.
I updated my implementation to use `initialize` to achieve trigger extension before training.
Could you review?

**Comment by wkentaro** [comment]
Now the training log looks like below:
```
epoch       iteration   main/loss   validation/main/loss  main/accuracy  validation/main/accuracy  elapsed_time
0           0                       2.27572                              0.1993                    0.801908
1           600         0.194542    0.0923036             0.94195        0.9711                    21.0456
2           1200        0.0733713   0.0769297             0.976917       0.977                     39.2025
3           1800        0.048582    0.071287              0.984517       0.9792                    59.7017
4           2400        0.0361622   0.0717561             0.9883         0.9795                    80.6741
5           3000        0.0296503   0.074036              0.990417       0.9785                    100.863
6           3600        0.0233984   0.083005              0.991867       0.9798                    120.973
7           4200        0.0209341   0.082974              0.993017       0.9785                    141.478
8           4800        0.0178388   0.0849701             0.994133       0.9803                    163.459
9           5400        0.0189349   0.071605              0.993983       0.982                     188.598
10          6000        0.0139158   0.113345              0.99555        0.9753                    210.196
11          6600        0.0150554   0.0873922             0.995133       0.9812                    231.432
12          7200        0.00938825  0.126652              0.997283       0.9743                    253.983
13          7800        0.0132886   0.0926085             0.996183       0.9802                    276.352
14          8400        0.0125203   0.0977689             0.996483       0.9803                    300.099
15          9000        0.0102308   0.100624              0.996983       0.9815                    325.268
```

**Comment by hvy** [comment]
Thank you for the fixes. I'll be taking another look!

File: examples/mnist/train_mnist.py [hunk 1]
```diff
@@ -77,7 +77,9 @@ def main():
     trainer = training.Trainer(updater, (args.epoch, 'epoch'), out=args.out)
 
     # Evaluate the model with the test dataset for each epoch
-    trainer.extend(extensions.Evaluator(test_iter, model, device=args.gpu))
+    evaluator = extensions.Evaluator(test_iter, model, device=args.gpu)
+    evaluator.initialize = evaluator.__call__  # for evaluation at epoch 0
```
Comment (hvy) [inline examples/mnist/train_mnist.py]: I am not sure about this interface. Can we not add an argument to the constructor instead?

**Comment by hvy** [comment]
Sorry for the late comment but would you mind taking a look? My question is how the user is supposed to control whether the extensions are invoked before training or not.

File: examples/mnist/train_mnist.py (hunk 1 above)
Comment (wkentaro) [reply examples/mnist/train_mnist.py]: I think there are 3 options:

1. `invoke_before_training=False/True` which implicitly overwrites `initialize` by `__call__`. This can cause unexpected behavior if user overwrites `Evaluator` and defines a custom `initialize` in it.
2. Passing `initialize=None/Extension.__call__/etc.` to the constructor and it overwrites the `initialize` method of the Extension if it is not `None`. This is fine, but it might be confusing that there is no way to overwrite other methods for example `__call__` and `finalize`. Also, it can't overwrite `initialize` method of extensions created by `extensions.make_extension`, for example `extensions.snapshot()`, so user must manually overwrite it.
3. Current implementation. It explicitly overwrites `initialize`, so it is less confusing. 

Currently, I'm conservative to adopt other implementations because of above problems caused by overwriting method in `__init__`.
What do you think?

**Comment by wkentaro** [comment]
Probably are you waiting for my comments? @hvy 
My comment is here https://github.com/chainer/chainer/pull/3511#discussion_r195986422.

File: examples/mnist/train_mnist.py (hunk 1 above)
Comment (hvy) [reply examples/mnist/train_mnist.py]: I think 1. is best but with some slight modifications. What do you think of the following?

- `call_before_training=False/True` passed to both `extend` and the constructors of the extensions. This flag is checked to decide whether to fire or not.
- Do not change the behavior of `initialize` since it's actual designed for a different purpose. Namely to update the trainer on resume.
- Make sure that the "0 iteration" invocations is not triggered on resume (check the iteration count).

**Comment by hvy** [comment]
Sorry for the delay. I wrote some comments and suggestion. Could you take a look at them  https://github.com/chainer/chainer/pull/3511#discussion_r ?

**Comment by knorth55** [comment]
@wkentaro Are you still working on this PR?

**Comment by wkentaro** [comment]
I will work on this in this week.

**Comment by wkentaro** [comment]
Sorry for very late. I was occupied by paper revisions.

I pushed new changes and now it looks much simpler. Could you check it? @hvy

**Comment by hvy** [comment]
Thanks for the changes although you've been busy. I'll try to take a look.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Is it possible to write tests for this feature?

**Comment by wkentaro** [comment]
> Is it possible to write tests for this feature?

I added a test in a74e0ad.

File: chainer/training/trainer.py [hunk 2]
```diff
@@ -308,6 +311,14 @@ def run(self, show_loop_exception_msg=True):
         reporter = self.reporter
         stop_trigger = self.stop_trigger
 
+        # call before training loop
```
[the remark this answers is not in the document]
Comment (wkentaro) [reply chainer/training/trainer.py]: I fixed it in 5a69154.

**Comment by wkentaro** [comment]
I'm still interested in this feature. Could anyone review this?

**Comment by toslunar** [comment]
Let me put the link to #2639.

**Comment by wkentaro** [comment]
You mean we should pass `call_before_training` argument to each Extension’s `__init__` instead of `Trainer.extend`? Why it is better?

**Comment by toslunar** [comment]
I've just understood reusing `initialize` is not a good idea.  I withdraw my previous comment.

**Comment by wkentaro** [comment]
Thanks. I added it.

File: chainer/training/trainer.py [hunk 3]
```diff
@@ -165,6 +166,10 @@ def __init__(self, updater, stop_trigger=None, out='result',
         for ext in extensions:
             self.extend(ext)
 
+    @property
+    def is_before_training(self):
```
Comment (hvy) [inline chainer/training/trainer.py]: Could you write a short documentation what this returns and that it can be used by extensions (for e.g. those who'd be interested in writing their own)?

**Comment by hvy** [comment]
Again, sorry for the delay. Could you take a look at my comment above and also resolve the conflicts? The changes otherwise LGTM!

**Comment by wkentaro** [comment]
Thanks for the review. I added the documentation and resolved the conflicts.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
LGTM, let me merge this long awaited PR. Thanks and sorry having it running for such a long time @wkentaro @toslunar .

**Comment by wkentaro** [comment]
Thank you for the reviews!
